File: friendly_introduction_to_number_theory/cpp/ex11_4_eulers_phi.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;

map<ll, ll> prime_factorize(ll n) {
  map<ll, ll> res;
  for (ll i = 2; i * i <= n; ++i) {
    while (n % i == 0) {
      ++res[i];
      n /= i;
    }
  }
  if (n > 1) ++res[n];
  return res;
}

ll phi(ll n) {
  map<ll, ll> factors = prime_factorize(n);
  ll res = n;
  for (auto p : factors) {
    res = res / p.first * (p.first - 1);
  }
  return res;
}

int main() {
  printf("phi(97)=%lld\n", phi(97));           // 96
  printf("phi(8800)=%lld\n", phi(8800));       // 3200
  printf("phi(1000000)=%lld\n", phi(1000000)); // 400000
}

File: friendly_introduction_to_number_theory/chapter11/python/ex11_4_eulers_phi.py
from collections import Counter


def prime_factorize(n):
    res = Counter()
    for i in range(2, int(n ** 0.5) + 1):
        while n % i == 0:
            res[i] += 1
            n //= i
    if n > 1:
        res[n] += 1
    return res


def phi(n):
    factors = prime_factorize(n)
    res = n
    for k in factors.keys():
        n = n // k * (k - 1)
    return res


print(f"phi(97)={phi(97)}")  # 96
print(f"phi(8800)={phi(8800)}")  # 8800
print(f"phi(1000000)={phi(1000000)}")   # 400000
